garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+
+    This class follows the BaseModel API and handles pickling automatically.
+    It provides both self.init_spec and self.parameters for pickling, which
+    are the arguments and weights of the Model. When unpickled, a new model
+    will be reconstructed with self.init_spec as arguments and load the weights
+    from self.parameters.
+
+    The target weights should be assigned to self._default_parameters during
+    pickling, so that the newly created model can check if target weights exist
+    or not.
+    """
+
+    def __init__(self):
+        self._name = type(self).__name__  # name default to class
+        self._inputs = None
+        self._outputs = None
+        self._default_parameters = None
+
+    @property
+    def init_spec(self):
+        """
+        Model specification.
+
+        Positional and named arguments (args and kwargs) which, when passed
+        to `__init__`, will construct this model.
+        """
+        pass
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    @property
+    def input(self):
+        """Tensor input of the Model."""
+        return self._inputs[0]
+
+    @property
+    def output(self):
+        """Tensor output of the Model."""
+        return self._outputs[0]
+
+    @property
+    def inputs(self):
+        """Tensor inputs of the Model."""
+        return self._inputs
+
+    @property
+    def outputs(self):
+        """Tensor outputs of the Model."""
+        return self._outputs
+
+    @property
+    def parameters(self):
+        """Parameters of the Model."""
+        raise NotImplementedError
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
+
+
+class TfModel(AutoPickableModel):
+    r"""
+    Model class for TensorFlow.
+
+    This class is a parent-child design.
+    The parent model contains the parameters, i.e. the structure, and
+    two child model are constructed with input_1 and input_2. Inputs and
+    outputs of the parent model will be binded with the default child, which
+    is the first child constructed.
+
+    When new child model is created, it reuses the parameter from the
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
+
+    The design is illustrated as the following:
```

Comment:
@krzentner has some concerns about this design, as expressed in https://github.com/rlworkgroup/garage/pull/507#discussion_r256672737

I am posting it here so we can close the other PR.

One alternative is for `build` to return a Graph object (with inputs and outputs), rather than storing the Graph internally and providing implicit access via setattr.

I don't think we can get around subgraph duplication, since TensorFlow basically forces that design. We will just have to choose how to represent it.

Reply:
Wow, PR's with shared commits are quite confusing.

I do think the alternative you describe is better, although I don't really understand the circumstances under which I would want to make multiple subgraphs from the same model. Can't I just make multiple model instances? Or does that not work?

Reply:
From my understanding, when you make multiple model instances, they will contain their own variables, bounded by their own `variable_scope`. Each of them will essentially be a subgraph. And then if we want those models to share the same variables, that's what this PR is trying to achieve, i.e. share `variable_scope`.

One example is DDPG, where the policy outputs an action, and passed to the Qfunction (QF) for optimizing policy loss. At the same time, samples from environment are passed to the same QF for optimizing QF loss. In this case, we have two subgraphs (policy loss and qf loss) but we share the same QF.

Reply:
i think i was able to explain this to KR last night